Fix: Errant GTID detection on the replicas when they set replication source

GuptaManan100 opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->
This PR adds errant GTID detection logic in `setReplicationSourceLocked` call. The intent is that we want to prevent any replica that has an errant GTID from even joining the replication stream. This is because we want to prevent this replica from sending semi-sync ACKs, which can land us into some really annoying and hard to deal with situations. 

To do the errant GTID detection, the tablets first query their own position and then ask the primary for its position. Then they compare the two to see if they have an errant GTID or not. If they do, they fail the RPC without setting the new source.

## Related Issue(s)

<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->
- Part of an effort to address the issue https://github.com/vitessio/vitess/issues/16724

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

shlomi-noach commented on go/mysql/replication/mysql56_gtid_set.go:
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
+
+	// Calculate the difference between the replica and primary GTID sets.
+	diffSet := replicaGTIDSet.Difference(primaryGTIDSet)
+	return diffSet.String(), nil
```
👍

shlomi-noach commented on go/mysql/replication/mysql56_gtid_set.go:
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
+		newSet[sid] = intervals
+	}
+	return newSet
+}
```
👍

shlomi-noach commented on go/mysql/replication/mysql56_gtid_set_test.go:
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
+			name:               "Single errant GTID",
+			replicaPosition:    "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
+			primaryPositionStr: "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-50,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-30",
+			errantGtidWanted:   "8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
```
👍

shlomi-noach commented on go/mysql/replication/mysql56_gtid_set_test.go:
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
+			name:               "Multiple errant GTID",
+			replicaPosition:    "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-32,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:3-35",
+			primaryPositionStr: "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-50,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-30,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
+			errantGtidWanted:   "8bc65cca-3fe4-11ed-bbfb-091034d48b3e:31-32,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:3-33:35",
```
👍

shlomi-noach commented on go/mysql/replication/mysql56_gtid_set_test.go:
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
+			name:       "Remove unknown UUID",
+			initialSet: "8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1:4-24",
+			uuid:       "8bc65c84-3fe4-11ed-a912-257f0fcde6c9",
+			wantSet:    "8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1:4-24",
```
👍

shlomi-noach commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -326,7 +323,7 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
 			// Reconnect to primary only for "NORMAL" keyspaces
 			params.Logger.Infof("Restore: starting replication at position %v", pos)
-			if err := tm.startReplication(context.Background(), pos, originalType); err != nil {
+			if err := tm.startReplication(ctx, pos, originalType); err != nil {
```
👍

shlomi-noach commented on go/vt/vttablet/tabletmanager/rpc_replication.go:
```diff
@@ -747,12 +754,32 @@ func (tm *TabletManager) setReplicationSourceLocked(ctx context.Context, parentA
+		// We will then compare our own position against it to verify that we don't
+		// have an errant GTID. If we find any GTID that we have, but the primary doesn't,
+		// we will not enter the replication graph and instead fail replication.
+		var primaryPosition string
+		primaryPosition, err = tm.tmc.PrimaryPosition(ctx, parent.Tablet)
```
for consistency and simplicity
```suggestion
		primaryPosition, err := tm.tmc.PrimaryPosition(ctx, parent.Tablet)
```

GuptaManan100 replied:
Fixed

shlomi-noach commented on go/vt/vttablet/tabletmanager/tm_init.go:
```diff
@@ -1003,21 +1010,47 @@ func (tm *TabletManager) initializeReplication(ctx context.Context, tabletType t
+	// We will then compare our own position against it to verify that we don't
+	// have an errant GTID. If we find any GTID that we have, but the primary doesn't,
+	// we will not enter the replication graph and instead fail replication.
+	var replicaPos replication.Position
+	replicaPos, err = tm.MysqlDaemon.PrimaryPosition(ctx)
```
For consistency and simplicity
```suggestion
	replicaPos, err := tm.MysqlDaemon.PrimaryPosition(ctx)
```

GuptaManan100 replied:
Fixed

shlomi-noach commented on go/vt/vttablet/tabletmanager/tm_init.go:
```diff
@@ -1003,21 +1010,47 @@ func (tm *TabletManager) initializeReplication(ctx context.Context, tabletType t
+		return "", err
 	}
+
+	var errantGTIDs string
+	errantGTIDs, err = replication.ErrantGTIDsOnReplica(replicaPos, primaryPosStr)
```
```suggestion
	errantGTIDs, err := replication.ErrantGTIDsOnReplica(replicaPos, primaryPosStr)
```

GuptaManan100 replied:
Fixed

shlomi-noach reviewed (approved):
LGTM (with comments). Are we convinced that `setReplicationSourceLocked` is the only path (other than `restore.go` which I've noticed is handled as well)?

deepthi commented on go/mysql/replication/mysql56_gtid_set.go:
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
 }
 
+// ErrantGTIDsOnReplica gets the errant GTIDs on the replica by comparing against the primary position and UUID.
+func ErrantGTIDsOnReplica(replicaPosition Position, primaryPositionStr string) (string, error) {
```
It would be more symmetric to pass in the primaryPosition as a `Position` instead of a string.

GuptaManan100 replied:
Fixed

deepthi reviewed (approved):
We should note in the release notes that for people running with vtop on k8s, this change means that the replica with broken replication will show as unready. There will be some kind of manual intervention needed to clean up these errant replica tablets, even if using the vtorc feature of marking them as drained.

GuptaManan100 commented:
Thank you for the reviews! I've addressed all review comments and added the release notes changes as requested. I will merge the PR once the tests are green.
